### PR TITLE
Filter:Triangle Centroid Filter

### PR DESCRIPTION
**Pull request by imikejackson**
## Naming Conventions

Naming of variables should descriptive where needed. Loop Control Variables can use `i` if warranted. Most of these conventions are enforced through the clang-tidy and clang-format configuration files.

- [x] Class and Structs are `UpperCamelCase`
- [x] Class private member variables are `m_UpperCamelCase`
- [x] Class methods are `lowerCamelCase`
- [x] Method arguments are `lowerCamelCase`
- [x] Normal variables are `lowerCamelCase`
- [x] Constants are `k_UpperCamelCase`
- [x] Global statics are `s_UpperCamelCase`
- [x] Free Functions are `UpperCamelCase`
- [x] Macros are `ALL_UPPER_SNAKE_CASE`
- [x] Unit test will test data output from the filter.
- [x] Reused strings should be constants in an anonymous namespace
- [x] If parallelization is used, proper use of the abstracted complex classes are used. Using TBB specifically in a filter should be frowned upon unless for a really good reason.

## Filter Checklist

- [x] Parameters should be generally broken down into "Input Parameters", "Required Data Objects", "Created Data Objects". There can be exceptions to this.
- [x] ChoicesParameter selections should be an enumeration defined in the filer header
- [x] Documentation copied from SIMPL Repo and updated (if necessary)
- [x] Parameter argument variables are k_CamelCase_Key
- [x] Parameter argument strings are lower_snake_case
```
static inline constexpr StringLiteral k_AlignmentType_Key = "alignment_type";
```

## Unit Testing
- [x] 1 Unit test to instantiate the filter with the default arguments.
- [x] 1 Unit test to test output from the filter against known exemplar set of data
- [x] 1 Unit test to test invalid input code paths that are specific to a filter. Don't test that a DataPath does not exist since that test is already performed as part of the SelectDataArrayAction.

## Code Cleanup
- [ ] No commented out code (rare exceptions to this is allowed..)
- [x] Filters should have both the Filter class and Algorithm class for anything beyond trivial needs
- [x] No API changes were made (or the changes have been approved)
- [x] No major design changes were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added license to new files (if any)
- [x] Added Python wrapping to new files (if any) as necessary
- [x] Added example pipelines that use the filter
- [x] Classes and methods are properly documented.

### CODE REVIEW

File: src/Plugins/ComplexCore/docs/TriangleCentroidFilter.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+None
+
+## Required Geometry ##
+
+Triangle
+
+## Required Objects ##
+
+None
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/docs/TriangleCentroidFilter.md:22]: Update the documentation to reflect the required objects

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/TriangleDihedralAngleFilter.cpp [hunk 2]
```diff
@@ -202,15 +199,15 @@ Result<> TriangleDihedralAngleFilter::executeImpl(DataStructure& dataStructure,
...
+  const TriangleGeom* triangleGeom = dataStructure.getDataAs<TriangleGeom>(pTriangleGeometryDataPath);
+  const AttributeMatrix* faceAttributeMatrix = triangleGeom->getFaceAttributeMatrix();
+  const DataPath dihedralAnglesArrayPath = pTriangleGeometryDataPath.createChildPath(faceAttributeMatrix->getName()).createChildPath(pMinDihedralAnglesName);
+  auto& dihedralAngles = dataStructure.getDataRefAs<Float64Array>(dihedralAnglesArrayPath);
 
   ParallelDataAlgorithm dataAlg;
-  dataAlg.setParallelizationEnabled(false);
-  dataAlg.setRange(0ULL, static_cast<size_t>(triangleGeom.getNumberOfFaces()));
-  dataAlg.execute(::CalculateDihedralAnglesImpl(*(triangleGeom.getVertices()), *(triangleGeom.getFaces()), dihedralAngles, shouldCancel));
+  dataAlg.setParallelizationEnabled(false); // <=== Why is this off?
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/TriangleDihedralAngleFilter.cpp]:
```suggestion
```
Not sure, probably was turned off to debug something and then it never got caught that it was running in serial

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/TriangleCentroidFilter.hpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+#include "ComplexCore/ComplexCore_export.hpp"
+
+#include "complex/Filter/FilterTraits.hpp"
+#include "complex/Filter/IFilter.hpp"
+
+namespace complex
+{
+/**
+ * @class TriangleCentroidFilter
+ * @brief This filter will ....
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/TriangleCentroidFilter.hpp]:
```suggestion
 * @brief This filter will calculate the centroids of a triangle list.
```

File: src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  for(auto element : faces)
+  {
+    underlyingStoreF->setValue(count++, element);
+  }
+  acuteTriangle.setFaceList(*facesList);
+  acuteTriangle.setVertices(*vertexList);
+
+  TriangleCentroidFilter filter;
+  Arguments args;
+  std::string centroidsArrayName = "Centroids";
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp]:
```suggestion
```

File: src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+using namespace complex::Constants;
+
+namespace
+{
+constexpr float64 k_MaxDifference = 0.00001;
+constexpr usize k_VertexTupleCount = 11;
+constexpr usize k_VertexCompCount = 3;
+constexpr usize k_FaceTupleCount = 4;
+constexpr usize k_FaceCompCount = 3;
+
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp:27]:
```suggestion
const std::string k_CentroidsArrayName = "Centroids";

```

File: src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  }
+  acuteTriangle.setFaceList(*facesList);
+  acuteTriangle.setVertices(*vertexList);
+
+  TriangleCentroidFilter filter;
+  Arguments args;
+  std::string centroidsArrayName = "Centroids";
+
+  DataPath geometryPath = dataStructure.getDataPathsForId(acuteTriangle.getId())[0];
+  DataPath triangleCentroidsDataPath = geometryPath.createChildPath(INodeGeometry2D::k_FaceDataName).createChildPath(centroidsArrayName);
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp]:
```suggestion
  DataPath triangleCentroidsDataPath = geometryPath.createChildPath(INodeGeometry2D::k_FaceDataName).createChildPath(k_CentroidsArrayName);
```

File: src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  auto preflightResult = filter.preflight(dataStructure, args);
+  COMPLEX_RESULT_REQUIRE_VALID(preflightResult.outputActions)
+
+  // Execute the filter and check the result
+  auto executeResult = filter.execute(dataStructure, args);
+  COMPLEX_RESULT_REQUIRE_VALID(executeResult.result)
+
+  std::vector<Point3Df> centroids = {{0.5F, 0.333333F, 0.0F}, {3.0F, 0.333333, 0.0F}, {6.66667, 0.0F, 0.0F}, {9.0F, 0.0F, 0.0F}};
+
+  Float32Array& calculatedCentroids = dataStructure.getDataRefAs<Float32Array>(triangleCentroidsDataPath);
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/test/TriangleCentroidTest.cpp]:
```suggestion
  auto& calculatedCentroids = dataStructure.getDataRefAs<Float32Array>(triangleCentroidsDataPath);
```

**Review by nyoungbq — APPROVED** [review]

File: src/Plugins/ComplexCore/docs/TriangleCentroidFilter.md (hunk 1 above)
Comment (imikejackson) [reply src/Plugins/ComplexCore/docs/TriangleCentroidFilter.md:22]: No required objects other than the Triangle Geometry

**Review by nyoungbq — APPROVED** [review]

**Pull request merged**